### PR TITLE
password gen, apex execution, package install, data import, hosted one-click option, yaml documentation

### PR DESCRIPTION
**Pull request by mshanemc**

### CODE REVIEW

**Comment by wadewegner** [comment]
Great pull request. There's a lot in here, so it'll take me a bit to go through it all. But thank you!

File: .gitignore [hunk 1]
```diff
@@ -38,4 +38,5 @@ public
 # Local resources
 tmp
 cert.pem
-key.pem
\ No newline at end of file
+key.pem
```
Comment (wadewegner) [inline .gitignore]: Do we need these .pem files? I don't think so.

File: lib/app-express.js [hunk 2]
```diff
@@ -1,11 +1,32 @@
...
 const consumerSecret = process.env.CONSUMERSECRET;
 
+const allowHostedOneClick = process.env.HOSTED_ONE_CLICK || false; // default for original flavor.
+let conn;
+
+// did this app or any of its peers already git init'd?  If so, there'll be a cached value for that
+if (allowHostedOneClick){
+  mjs.get('conn', (err, cache) => {
+    if (err){
+      console.log(err);
```
Comment (wadewegner) [inline lib/app-express.js]: We should include details like `console.log('Error getting cache info', err);`.

File: lib/app-express.js [hunk 3]
```diff
@@ -1,11 +1,32 @@
...
+    }
+    if (cache){
+      console.log('loading conn from cache');
```
Comment (wadewegner) [inline lib/app-express.js]: Unless it's absolutely necessary, I'd rather not log information. It'll end up in the Heroku logs.

File: lib/app-express.js [hunk 4]
```diff
@@ -16,6 +37,48 @@ module.exports = function (app) {
     return next();
   });
 
+  app.get('/hostedOneClick', (req, res) => {
```
Comment (wadewegner) [inline lib/app-express.js]: I'm finding this pretty hard to follow. I wonder, should we create some single purpose functions for some in a different file?

File: lib/app-express.js [hunk 5]
```diff
@@ -16,6 +37,48 @@ module.exports = function (app) {
...
+    // feature not active--send you to the homepage
+    console.log('oneClick requested');
+    const template = req.query.template;
+    if (!allowHostedOneClick){
+      console.log('OneClick deploy not allowed');
+      res.render('pages/customError', {
```
Comment (wadewegner) [inline lib/app-express.js]: Instead of another error page, what if we add the custom error support into the main error page?

File: lib/app-express.js [hunk 6]
```diff
@@ -106,11 +169,31 @@ module.exports = function (app) {
...
+
+    return res.redirect(uri);
+  });
+
   app.get('/logout', (req, res) => {
     const access_token = req.cookies.access_token;
     const instance_url = req.cookies.instance_url;
 
-    const conn = new jsforce.Connection({
+    conn = new jsforce.Connection({
```
Comment (wadewegner) [inline lib/app-express.js]: Should be `const`.

File: lib/app-express.js [hunk 7]
```diff
@@ -130,6 +213,8 @@ module.exports = function (app) {
     const authorizationCode = req.param('code');
     const template = req.param('state');
 
+    console.log(`template is ${template}`);
```
Comment (wadewegner) [inline lib/app-express.js]: Let's remove unnecessary logging

File: lib/app-express.js [hunk 8]
```diff
@@ -138,7 +223,7 @@ module.exports = function (app) {
     }, (error, payload) => {
 
       try {
-
+        console.log(payload);
```
Comment (wadewegner) [inline lib/app-express.js]: Remove

File: lib/app-express.js [hunk 9]
```diff
@@ -149,27 +234,59 @@ module.exports = function (app) {
       }
 
       // check to see if org is a dev hub
-      const conn = new jsforce.Connection({
+      let connLocal = new jsforce.Connection({
```
Comment (wadewegner) [inline lib/app-express.js]: Should be `const`

Comment (wadewegner) [reply lib/app-express.js]: Also, why renamed?

File: lib/app-express.js [hunk 10]
```diff
@@ -149,27 +234,59 @@ module.exports = function (app) {
...
         instanceUrl: payload.instance_url,
         accessToken: payload.access_token
       });
 
-      conn.identity((err, identity) => {
+      connLocal.identity((err, identity) => {
         if (err) {
           return console.error(err);
         }
+        console.log(identity);
```
Comment (wadewegner) [inline lib/app-express.js]: This will put info into logs that shouldn't be persisted there

File: lib/app-router.js [hunk 11]
```diff
@@ -84,6 +87,40 @@ module.exports = function (router) {
 
         break;
 
+      case 'apex':
+        let message = 'Apex Scripts run.';
+        async.eachSeries(param, function iteratee(apex, callback) {
```
Comment (wadewegner) [inline lib/app-router.js]: What affect do this have on the ajax calls? I had been running most things synchronously so that the output works properly.

File: lib/app-router.js [hunk 12]
```diff
@@ -117,6 +154,7 @@ module.exports = function (router) {
               message: `Error: ${commandErr}.`
             });
           } else {
+            console.log(commandResult);
```
Comment (wadewegner) [inline lib/app-router.js]: Remove

File: lib/app-router.js [hunk 13]
```diff
@@ -84,6 +87,40 @@ module.exports = function (router) {
...
+          script = `${startingDirectory}cd ${directory};export FORCE_SHOW_SPINNER=;sfdx force:apex:execute -f ${apex}`;
+
+          commands.run(command, script, (commandResult, commandErr) => {
+            if (commandErr) {
+              console.log(commandErr);
+              callback(commandErr);
+
+            } else {
+              console.log(commandResult);
+              // doesn't seem to be returning until after everything else runs.  fix UI output
```
Comment (wadewegner) [inline lib/app-router.js]: Yeah, that's because it's async. I think we should change the behavior if it's not outputting correctly.

File: lib/app-router.js [hunk 14]
```diff
@@ -84,6 +87,40 @@ module.exports = function (router) {
...
+              message = `${message}
+              \t${apex}`;
+              callback();
+            }
+          });
+        }, function(err){
+          if (err){
```
Comment (wadewegner) [inline lib/app-router.js]: Is this a new pattern for handling errors? If so, let's create a function out of it and use it consistently.

File: scripts/deploying.js [hunk 15]
```diff
@@ -1,3 +1,5 @@
+/* globals $,document, jsyaml */
```
Comment (wadewegner) [inline scripts/deploying.js]: Out of curiosity, what's the purpose?

File: views/pages/customError.ejs [hunk 16]
```diff
@@ -0,0 +1,15 @@
+<!-- views/pages/error.ejs -->
```
Comment (wadewegner) [inline views/pages/customError.ejs]: Let's add this functionality into the main error page

File: lib/app-express.js (hunk 4 above)
Comment (wadewegner) [reply lib/app-express.js]: What happens if `hostedOneClick?template=init` is called more than once?

File: lib/app-express.js (hunk 2 above)
Comment (mshanemc) [reply lib/app-express.js]: agreed, done.

File: lib/app-express.js (hunk 4 above)
Comment (mshanemc) [reply lib/app-express.js]: these get fired if you're initing and it's already been done.  

``` javascript
if (template === 'init'){
  res.render('pages/customError', {
    customError : 'You\'re setting the template to init, but the app is already got its credentials loaded'
  });
}
```

File: lib/app-express.js (hunk 5 above)
Comment (mshanemc) [reply lib/app-express.js]: Yeah--I was trying not to make huge changes to your existing stuff, but I agree that makes more sense.  Yours was moving towards extraneous pages for known errors like "not a hub".

File: lib/app-express.js (hunk 7 above)
Comment (mshanemc) [reply lib/app-express.js]: agreed, done.

File: lib/app-express.js (hunk 8 above)
Comment (mshanemc) [reply lib/app-express.js]: agreed, done

File: lib/app-express.js (hunk 9 above)
Comment (mshanemc) [reply lib/app-express.js]: I stole conn and made it a global for the shared hub option.  connLocal is the original version (BYOhub)

File: lib/app-express.js (hunk 10 above)
Comment (mshanemc) [reply lib/app-express.js]: agreed, done.

File: lib/app-router.js (hunk 13 above)
Comment (mshanemc) [reply lib/app-router.js]: actually, this was fixed by adding in the async from the previous comment--I just forgot to take out my comment.  

Because the yaml lets your have multiple apex scripts to run (we're going to have the same with data import and packages!) we need to have the run in the order specified by the user.  Because the ajax calls were async, the loop needed to be async before returning.  

If that doesn't make sense in theory, it certainly worked in practice.

File: lib/app-router.js (hunk 14 above)
Comment (mshanemc) [reply lib/app-router.js]: it's only necessary for the async loop--basically, if anything in the iteration throws an error, we want to catch and pass back.

So yes, when I add package(s) and data import(s) I'll use the same pattern.

File: scripts/deploying.js (hunk 15 above)
Comment (mshanemc) [reply scripts/deploying.js]: makes eslint not gripe about undefined so it can focus on my real problems.

File: lib/app-router.js (hunk 12 above)
Comment (mshanemc) [reply lib/app-router.js]: agreed, done.  aside: Any thoughts to what you'd want the output from apex scripts to look like?  Right now, the UI just informs that each one ran.

File: views/pages/customError.ejs (hunk 16 above)
Comment (mshanemc) [reply views/pages/customError.ejs]: done.  Redid your token error handling to use it, and remove the route for /error

File: .gitignore (hunk 1 above)
Comment (mshanemc) [reply .gitignore]: I don't think so either--I never had them.  So let's leave them in the .gitignore?

File: lib/app-express.js (hunk 3 above)
Comment (mshanemc) [reply lib/app-express.js]: done.

File: lib/app-express.js (hunk 4 above)
Comment (mshanemc) [reply lib/app-express.js]: Here's an important question that just came to me.  

Why does your version of the the app (BYOhub) need oauth key/secret stored in it, if users are using their own hubs to sign in and generate orgs off of?

File: lib/app-express.js (hunk 6 above)
Comment (mshanemc) [reply lib/app-express.js]: fixed, as a new local const.  I'm using *conn* as the global for the hosted hub (set from init or from cache).

File: lib/app-router.js (hunk 11 above)
Comment (mshanemc) [reply lib/app-router.js]: see next comment.

**Comment by mshanemc** [comment]
the high-level discussion:

Right now, it's set up so that
- / is the front door, like your current BYOhub version.
- /hostedOneClick is like the secret side entrance that you invite your friends to, OR, if you have unlimited scratch orgs, you can send everyone there with your links/buttons/etc

An alternative would be to modify **index.ejs** so that, if the **allowHostedOneClick** env is true, the page presents both options.

The challenge that doesn't address if people don't have a hub OR a template.  So choose.ejs would also need to handle hosted-scratches.

File: lib/app-express.js (hunk 4 above)
Comment (wadewegner) [reply lib/app-express.js]: Because they're using my Connected App.

**Comment by wadewegner** [comment]
I'm going to start cherry picking updates from here but won't take the dev hub approach. Let's work on it post-DF17.
